### PR TITLE
MLE and L1 loss functions

### PR DESCRIPTION
**Pull request by k78ma**
Addresses #76 :)

### CODE REVIEW

File: python/mlx/nn/losses.py [hunk 1]
```diff
@@ -2,7 +2,16 @@
...
     score = mx.take_along_axis(logits, targets[..., None], axis).squeeze(-1)
     return mx.logsumexp(logits, axis=axis) - score
+
+def MLE_loss(logits: mx.array, targets: mx.array, axis: int = -1):
+    logits = logits- mx.max(logits, axis=axis, keepdims=True)
+    log_probs = mx.log(mx.softmax(logits, axis=axis))
+    score = mx.take_along_axis(log_probs, targets[..., None], axis).squeeze(-1)
+    return -mx.mean(score)
+
+def L1_loss(predictions: mx.array, targets: mx.array):
```
Comment (awni) [inline python/mlx/nn/losses.py:15]: I think this should be lower case `l1_loss` to fit the style.

File: python/mlx/nn/losses.py [hunk 2]
```diff
@@ -2,7 +2,16 @@
 
 import mlx.core as mx
 
-
 def cross_entropy(logits: mx.array, targets: mx.array, axis: int = -1):
     score = mx.take_along_axis(logits, targets[..., None], axis).squeeze(-1)
     return mx.logsumexp(logits, axis=axis) - score
+
+def MLE_loss(logits: mx.array, targets: mx.array, axis: int = -1):
```
Comment (awni) [inline python/mlx/nn/losses.py:9]: What does `MLE` stand for? This loss looks the same as the `cross_entropy` except it does a mean at the end? I think it would make more sense to just add a reduce option to the `cross_entropy` loss.

**Review by awni — CHANGES_REQUESTED** [review]
Thanks! I left a couple of comments. Also could you add a docstring for the losses? Then we can actually add them into the docs :).

File: python/mlx/nn/losses.py (hunk 2 above)
Comment (awni) [reply python/mlx/nn/losses.py:9]: I like the way PyTorch does reductions, maybe we follow that with the `reduction` argument?

**Comment by k78ma** [comment]
- Reduction argument added to `cross_entropy()` following PyTorch
- Docstrings added
- `l1_loss` name change
- Removed redundant MLE loss

**Comment by k78ma** [comment]
Forgot to change `l1_loss` name in tests whoops

File: python/mlx/nn/losses.py [hunk 3]
```diff
@@ -2,7 +2,46 @@
...
+                                    'mean': the sum of the output will be divided by the number of elements in the output.
+                                    'sum': the output will be summed. 
+                                    Defaults to 'mean'.
+
+    Returns:
+        mx.array: The computed cross entropy loss.
+    """
+    logits = logits - mx.max(logits, axis=axis, keepdims=True)
+    log_probs = mx.log(mx.softmax(logits, axis=axis))
+    score = mx.take_along_axis(log_probs, targets[..., None], axis).squeeze(-1)
```
Comment (awni) [inline python/mlx/nn/losses.py:24]: Would you mind reverting this to the old implementation, it is faster and numerically better as it avoids the explicit `softmax`. Also FYI for future reference `mx.softmax` is numerically stable you don't need to subtract the max for it.

File: python/mlx/nn/losses.py [hunk 4]
```diff
@@ -2,7 +2,46 @@
 
 import mlx.core as mx
 
+def cross_entropy(logits: mx.array, targets: mx.array, axis: int = -1, reduction: str = 'mean'):
```
Comment (awni) [inline python/mlx/nn/losses.py:5]: could you default `reduction` to `'none'` for now? I know it's different than PyTorch but we have downstream code which depends on this function that will otherwise break.

File: python/tests/test_nn.py [hunk 5]
```diff
@@ -21,8 +20,27 @@ def test_linear(self):
     def test_cross_entropy(self):
         logits = mx.array([[0.0, -float("inf")], [-float("inf"), 0.0]])
         targets = mx.array([0, 1])
-        losses = nn.losses.cross_entropy(logits, targets)
-        self.assertTrue(mx.array_equal(losses, mx.zeros((2,))))
+
+        # Test with reduction 'none'
```
Comment (awni) [inline python/tests/test_nn.py:24]: 👍

**Review by awni — COMMENTED** [review]
Thank you for updating this and adding the tests + docstring.

Just a couple more requests, then we can merge this!

**Comment by k78ma** [comment]
Should be good now I think?
- Default set to `none`
- Reverted to old implementation for loss calculation w/o explicit softmax

**Review by awni — APPROVED** [review]
Brilliant, thanks!
